### PR TITLE
coord: Check role validity after async work

### LINKED ISSUE #21639: parallel-workload: thread 'coordinator' panicked at 'catalog out of sync', src/adapter/src/catalog.rs:715:34
### What version of Materialize are you using?

090f121652

### What is the issue?

```
parallel-workload-materialized-1  | thread 'coordinator' panicked at 'catalog out of sync', src/adapter/src/catalog.rs:715:34
parallel-workload-materialized-1  | stack backtrace:
parallel-workload-materialized-1  |    0: rust_begin_unwind
parallel-workload-materialized-1  |              at ./rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/panicking.rs:593:5
parallel-workload-materialized-1  |    1: core::panicking::panic_fmt
parallel-workload-materialized-1  |              at ./rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/panicking.rs:67:14
parallel-workload-materialized-1  |    2: core::panicking::panic_display
parallel-workload-materialized-1  |              at ./rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/panicking.rs:150:5
parallel-workload-materialized-1  |    3: core::panicking::panic_str
parallel-workload-materialized-1  |              at ./rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/panicking.rs:134:5
parallel-workload-materialized-1  |    4: core::option::expect_failed
parallel-workload-materialized-1  |              at ./rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/option.rs:1952:5
parallel-workload-materialized-1  |    5: <core::option::Option<&mz_adapter::catalog::Role>>::expect
parallel-workload-materialized-1  |              at ./rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/option.rs:898:21
parallel-workload-materialized-1  |    6: <mz_adapter::catalog::CatalogState>::get_role
parallel-workload-materialized-1  |              at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-03cef5cfcfa9aa4d4-1/materialize/tests/src/adapter/src/catalog.rs:715:34
parallel-workload-materialized-1  |    7: <mz_adapter::catalog::ConnCatalog as mz_sql::catalog::SessionCatalog>::get_role
parallel-workload-materialized-1  |              at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-03cef5cfcfa9aa4d4-1/materialize/tests/src/adapter/src/catalog.rs:8699:9
parallel-workload-materialized-1  |    8: mz_adapter::rbac::check_plan::<mz_adapter::catalog::ConnCatalog>::{closure#0}
parallel-workload-materialized-1  |              at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-03cef5cfcfa9aa4d4-1/materialize/tests/src/adapter/src/rbac.rs:267:28
parallel-workload-materialized-1  |    9: core::iter::adapters::map::map_fold::<&mz_repr::role_id::RoleId, alloc::string::String, (), mz_adapter::rbac::check_plan<mz_adapter::catalog::ConnCatalog>::{closure#0}, core::iter::traits::iterator::Iterator::for_each::call<alloc::string::String, <alloc::vec::Vec<alloc::string::String>>::extend_trusted<core::iter::adapters::map::Map<alloc::vec::into_iter::IntoIter<&mz_repr::role_id::RoleId>, mz_adapter::rbac::check_plan<mz_adapter::catalog::ConnCatalog>::{closure#0}>>::{closure#0}>::{closure#0}>::{closure#0}
parallel-workload-materialized-1  |              at ./rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/iter/adapters/map.rs:84:28
parallel-workload-materialized-1  |   10: <alloc::vec::into_iter::IntoIter<&mz_repr::role_id::RoleId> as core::iter::traits::iterator::Iterator>::fold::<(), core::iter::adapters::map::map_fold<&mz_repr::role_id::RoleId, alloc::string::String, (), mz_adapter::rbac::check_plan<mz_adapter::catalog::ConnCatalog>::{closure#0}, core::iter::traits::iterator::Iterator::for_each::call<alloc::string::String, <alloc::vec::Vec<alloc::string::String>>::extend_trusted<core::iter::adapters::map::Map<alloc::vec::into_iter::IntoIter<&mz_repr::role_id::RoleId>, mz_adapter::rbac::check_plan<mz_adapter::catalog::ConnCatalog>::{closure#0}>>::{closure#0}>::{closure#0}>::{closure#0}>
parallel-workload-materialized-1  |              at ./rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/iter/traits/iterator.rs:2481:21
parallel-workload-materialized-1  |   11: <core::iter::adapters::map::Map<alloc::vec::into_iter::IntoIter<&mz_repr::role_id::RoleId>, mz_adapter::rbac::check_plan<mz_adapter::catalog::ConnCatalog>::{closure#0}> as core::iter::traits::iterator::Iterator>::fold::<(), core::iter::traits::iterator::Iterator::for_each::call<alloc::string::String, <alloc::vec::Vec<alloc::string::String>>::extend_trusted<core::iter::adapters::map::Map<alloc::vec::into_iter::IntoIter<&mz_repr::role_id::RoleId>, mz_adapter::rbac::check_plan<mz_adapter::catalog::ConnCatalog>::{closure#0}>>::{closure#0}>::{closure#0}>
parallel-workload-materialized-1  |              at ./rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/iter/adapters/map.rs:124:9
parallel-workload-materialized-1  |   12: <core::iter::adapters::map::Map<alloc::vec::into_iter::IntoIter<&mz_repr::role_id::RoleId>, mz_adapter::rbac::check_plan<mz_adapter::catalog::ConnCatalog>::{closure#0}> as core::iter::traits::iterator::Iterator>::for_each::<<alloc::vec::Vec<alloc::string::String>>::extend_trusted<core::iter::adapters::map::Map<alloc::vec::into_iter::IntoIter<&mz_repr::role_id::RoleId>, mz_adapter::rbac::check_plan<mz_adapter::catalog::ConnCatalog>::{closure#0}>>::{closure#0}>
parallel-workload-materialized-1  |              at ./rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/iter/traits/iterator.rs:856:9
parallel-workload-materialized-1  |   13: <alloc::vec::Vec<alloc::string::String>>::extend_trusted::<core::iter::adapters::map::Map<alloc::vec::into_iter::IntoIter<&mz_repr::role_id::RoleId>, mz_adapter::rbac::check_plan<mz_adapter::catalog::ConnCatalog>::{closure#0}>>
parallel-workload-materialized-1  |              at ./rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/alloc/src/vec/mod.rs:2828:17
parallel-workload-materialized-1  |   14: <alloc::vec::Vec<alloc::string::String> as alloc::vec::spec_extend::SpecExtend<alloc::string::String, core::iter::adapters::map::Map<alloc::vec::into_iter::IntoIter<&mz_repr::role_id::RoleId>, mz_adapter::rbac::check_plan<mz_adapter::catalog::ConnCatalog>::{closure#0}>>>::spec_extend
parallel-workload-materialized-1  |              at ./rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/alloc/src/vec/spec_extend.rs:26:9
parallel-workload-materialized-1  |   15: <alloc::vec::Vec<alloc::string::String> as alloc::vec::spec_from_iter_nested::SpecFromIterNested<alloc::string::String, core::iter::adapters::map::Map<alloc::vec::into_iter::IntoIter<&mz_repr::role_id::RoleId>, mz_adapter::rbac::check_plan<mz_adapter::catalog::ConnCatalog>::{closure#0}>>>::from_iter
parallel-workload-materialized-1  |              at ./rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/alloc/src/vec/spec_from_iter_nested.rs:62:9
parallel-workload-materialized-1  |   16: <alloc::vec::Vec<alloc::string::String> as alloc::vec::spec_from_iter::SpecFromIter<alloc::string::String, core::iter::adapters::map::Map<alloc::vec::into_iter::IntoIter<&mz_repr::role_id::RoleId>, mz_adapter::rbac::check_plan<mz_adapter::catalog::ConnCatalog>::{closure#0}>>>::from_iter
parallel-workload-materialized-1  |              at ./rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/alloc/src/vec/in_place_collect.rs:167:20
parallel-workload-materialized-1  |   17: <alloc::vec::Vec<alloc::string::String> as core::iter::traits::collect::FromIterator<alloc::string::String>>::from_iter::<core::iter::adapters::map::Map<alloc::vec::into_iter::IntoIter<&mz_repr::role_id::RoleId>, mz_adapter::rbac::check_plan<mz_adapter::catalog::ConnCatalog>::{closure#0}>>
parallel-workload-materialized-1  |              at ./rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/alloc/src/vec/mod.rs:2696:9
parallel-workload-materialized-1  |   18: <core::iter::adapters::map::Map<alloc::vec::into_iter::IntoIter<&mz_repr::role_id::RoleId>, mz_adapter::rbac::check_plan<mz_adapter::catalog::ConnCatalog>::{closure#0}> as core::iter::traits::iterator::Iterator>::collect::<alloc::vec::Vec<alloc::string::String>>
parallel-workload-materialized-1  |              at ./rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/iter/traits/iterator.rs:1895:9
parallel-workload-materialized-1  |   19: mz_adapter::rbac::check_plan::<mz_adapter::catalog::ConnCatalog>
parallel-workload-materialized-1  |              at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-03cef5cfcfa9aa4d4-1/materialize/tests/src/adapter/src/rbac.rs:268:14
parallel-workload-materialized-1  |   20: <mz_adapter::coord::Coordinator>::sequence_plan::{closure#0}::{closure#0}
parallel-workload-materialized-1  |              at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-03cef5cfcfa9aa4d4-1/materialize/tests/src/adapter/src/coord/sequencer.rs:99:25
parallel-workload-materialized-1  |   21: <tracing::instrument::Instrumented<<mz_adapter::coord::Coordinator>::sequence_plan::{closure#0}::{closure#0}> as core::future::future::Future>::poll
parallel-workload-materialized-1  |              at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
parallel-workload-materialized-1  |   22: <mz_adapter::coord::Coordinator>::sequence_plan::{closure#0}
parallel-workload-materialized-1  |              at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-03cef5cfcfa9aa4d4-1/materialize/tests/src/adapter/src/coord/sequencer.rs:63:5
parallel-workload-materialized-1  |   23: <mz_adapter::coord::Coordinator>::handle_execute_inner::{closure#0}::{closure#0}
parallel-workload-materialized-1  |              at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-03cef5cfcfa9aa4d4-1/materialize/tests/src/adapter/src/coord/command_handler.rs:654:73
parallel-workload-materialized-1  |   24: <tracing::instrument::Instrumented<<mz_adapter::coord::Coordinator>::handle_execute_inner::{closure#0}::{closure#0}> as core::future::future::Future>::poll
parallel-workload-materialized-1  |              at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
parallel-workload-materialized-1  |   25: <mz_adapter::coord::Coordinator>::handle_execute_inner::{closure#0}
parallel-workload-materialized-1  |              at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-03cef5cfcfa9aa4d4-1/materialize/tests/src/adapter/src/coord/command_handler.rs:426:5
parallel-workload-materialized-1  |   26: <mz_adapter::coord::Coordinator>::handle_execute::{closure#0}::{closure#0}
parallel-workload-materialized-1  |              at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-03cef5cfcfa9aa4d4-1/materialize/tests/src/adapter/src/coord/command_handler.rs:423:54
parallel-workload-materialized-1  |   27: <tracing::instrument::Instrumented<<mz_adapter::coord::Coordinator>::handle_execute::{closure#0}::{closure#0}> as core::future::future::Future>::poll
parallel-workload-materialized-1  |              at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
parallel-workload-materialized-1  |   28: <mz_adapter::coord::Coordinator>::handle_execute::{closure#0}
parallel-workload-materialized-1  |              at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-03cef5cfcfa9aa4d4-1/materialize/tests/src/adapter/src/coord/command_handler.rs:324:5
parallel-workload-materialized-1  |   29: <tracing::instrument::Instrumented<<mz_adapter::coord::Coordinator>::handle_execute::{closure#0}> as core::future::future::Future>::poll
parallel-workload-materialized-1  |              at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
parallel-workload-materialized-1  |   30: <mz_adapter::coord::Coordinator>::handle_command::{closure#0}
parallel-workload-materialized-1  |              at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-03cef5cfcfa9aa4d4-1/materialize/tests/src/adapter/src/coord/command_handler.rs:100:22
parallel-workload-materialized-1  |   31: <mz_adapter::coord::Coordinator>::message_command::{closure#0}::{closure#0}
parallel-workload-materialized-1  |              at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-03cef5cfcfa9aa4d4-1/materialize/tests/src/adapter/src/coord/message_handler.rs:239:34
parallel-workload-materialized-1  |   32: <tracing::instrument::Instrumented<<mz_adapter::coord::Coordinator>::message_command::{closure#0}::{closure#0}> as core::future::future::Future>::poll
parallel-workload-materialized-1  |              at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
parallel-workload-materialized-1  |   33: <mz_adapter::coord::Coordinator>::message_command::{closure#0}
parallel-workload-materialized-1  |              at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-03cef5cfcfa9aa4d4-1/materialize/tests/src/adapter/src/coord/message_handler.rs:236:5
parallel-workload-materialized-1  |   34: <mz_adapter::coord::Coordinator>::handle_message::{closure#0}
parallel-workload-materialized-1  |              at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-03cef5cfcfa9aa4d4-1/materialize/tests/src/adapter/src/coord/message_handler.rs:44:64
parallel-workload-materialized-1  |   35: <tracing::instrument::Instrumented<<mz_adapter::coord::Coordinator>::handle_message::{closure#0}> as core::future::future::Future>::poll
parallel-workload-materialized-1  |              at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
parallel-workload-materialized-1  |   36: <mz_ore::metrics::WallTimeFuture<tracing::instrument::Instrumented<<mz_adapter::coord::Coordinator>::handle_message::{closure#0}>, prometheus::histogram::Histogram> as core::future::future::Future>::poll
parallel-workload-materialized-1  |              at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-03cef5cfcfa9aa4d4-1/materialize/tests/src/ore/src/metrics.rs:592:28
parallel-workload-materialized-1  |   37: <mz_adapter::coord::Coordinator>::serve::{closure#0}
parallel-workload-materialized-1  |              at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-03cef5cfcfa9aa4d4-1/materialize/tests/src/adapter/src/coord.rs:1912:18
parallel-workload-materialized-1  |   38: <tokio::runtime::park::CachedParkThread>::block_on::<<mz_adapter::coord::Coordinator>::serve::{closure#0}>::{closure#0}
parallel-workload-materialized-1  |              at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.27.0/src/runtime/park.rs:283:63
parallel-workload-materialized-1  |   39: tokio::runtime::coop::with_budget::<core::task::poll::Poll<()>, <tokio::runtime::park::CachedParkThread>::block_on<<mz_adapter::coord::Coordinator>::serve::{closure#0}>::{closure#0}>
parallel-workload-materialized-1  |              at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.27.0/src/runtime/coop.rs:107:5
parallel-workload-materialized-1  |   40: tokio::runtime::coop::budget::<core::task::poll::Poll<()>, <tokio::runtime::park::CachedParkThread>::block_on<<mz_adapter::coord::Coordinator>::serve::{closure#0}>::{closure#0}>
parallel-workload-materialized-1  |              at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.27.0/src/runtime/coop.rs:73:5
parallel-workload-materialized-1  |   41: <tokio::runtime::park::CachedParkThread>::block_on::<<mz_adapter::coord::Coordinator>::serve::{closure#0}>
parallel-workload-materialized-1  |              at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.27.0/src/runtime/park.rs:283:31
parallel-workload-materialized-1  |   42: <tokio::runtime::context::BlockingRegionGuard>::block_on::<<mz_adapter::coord::Coordinator>::serve::{closure#0}>
parallel-workload-materialized-1  |              at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.27.0/src/runtime/context.rs:315:13
parallel-workload-materialized-1  |   43: <tokio::runtime::handle::Handle>::block_on::<<mz_adapter::coord::Coordinator>::serve::{closure#0}>
parallel-workload-materialized-1  |              at ./cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.27.0/src/runtime/handle.rs:264:9
parallel-workload-materialized-1  |   44: mz_adapter::coord::serve::{closure#0}::{closure#1}
parallel-workload-materialized-1  |              at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-03cef5cfcfa9aa4d4-1/materialize/tests/src/adapter/src/coord.rs:2151:17
parallel-workload-materialized-1  | note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```
Seen in parallel-workload with cancellations: https://buildkite.com/materialize/nightlies/builds/3850#018a6cd6-6d3e-4836-9c40-f06d144e0f82

@jkosh44 This panic seems to come from https://github.com/MaterializeInc/materialize/commit/9f555d09ad2ff93f6e68c8fe91462de54d9b7c69

ci-regexp: thread 'coordinator' panicked at 'catalog out of sync'

### PR DESCRIPTION
**Pull request by jkosh44**
This commit improves the validity check of async coord work by checking that the session role IDs still exist. Without this check, the coordinator may assume that the roles still exist and panic after trying to get them from the catalog.

Fixes #21639

### Motivation
This PR fixes a recognized bug.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Fix panics caused by concurrent role drops.

### CODE REVIEW

**Review by def- — APPROVED** [review]
Great, thanks for fixing so quickly!

Is it possible to write a deterministic test for this panic?

I'm triggering nightly to check that this fixes the issue in CI too.

**Comment by jkosh44** [comment]
> Is it possible to write a deterministic test for this panic?

I don't think we have the proper test infrastructure to write deterministic tests with asynchronous Coordinator code. Is this something we could use fail points for?

Also looking at the stack trace, I'm still very unsure how this error happens and how my fix prevents it.

**Comment by jkosh44** [comment]
So this isn't the direct cause of https://github.com/MaterializeInc/materialize/issues/21639, but I still think it's a different issue that we should fix? I'm not actually sure. Is it OK for a session's role to be dropped mid query? Probably not, but I'm not sure.

File: src/adapter/src/coord.rs [hunk 1]
```diff
@@ -473,6 +474,15 @@ impl PlanValidity {
...
             }
         }
+        if catalog
+            .try_get_role(&self.role_metadata.current_role)
+            .is_none()
+            || catalog
+                .try_get_role(&self.role_metadata.session_role)
+                .is_none()
+        {
+            return Err(AdapterError::ChangedPlan);
```
Comment (andrewrodriguez-m) [inline src/adapter/src/coord.rs:484]: should this be AdapterError::Unauthorized ?

**Review by andrewrodriguez-m — APPROVED** [review]

File: src/adapter/src/coord.rs (hunk 1 above)
Comment (jkosh44) [reply src/adapter/src/coord.rs]: Yeah, that probably works better, just updated.

**Comment by jkosh44** [comment]
Even though this may not be strictly necessary, it's still safer to have. So I'm going to go ahead and merge this.

**Pull request merged**